yt/visualization/accumulators.py

```diff
@@ -0,0 +1,290 @@
+        The path to be integrated along
+
+    field_vals : YTArray
+        An array containing the components of the vector field to be
+        integrated. The values are sampled at the starting point of
+        each path segment as well as the endpoint for the last segment
+
+    Returns
+    -------
+        accum : YTArray
+            The cumulative value of the field integral at each path
+            segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.dot doesn't combine units correctly and there does not appear to be a
+    # unyt implementation of dot, so units will be handled manually
+    # for the time being
+    accum = np.cumsum(np.diag(np.dot(field_vals[:-1].d, (path[1:].d - path[:-1].d).T)))
+    accum = YTArray(accum, field_vals.units * path.units)
+    return accum
+
+
+def _accumulate_scalar_field(p, field_vals):
+    r"""
+    This function integrates a scalar field along a path. It uses a
+    similar method to that in _accumulate_vector_field, but the integral
+    is now:
+
+    ..math::
+
+        I = \int_C \phi(x1,x2,...,xn)d\vec{r}
+
+    Parameters
+    ----------
+    p : YTArray
+        The path to be integrated along
+
+    field_vals : YTArray
+        An array containing the values of the scalar field to be
+        integrated at the location of the starting point of each
+        path segment as well as the endpoint for the last segment
+
+    Returns
+    -------
+    accum : YTArray
+        The cumulative value of the field integral at each path
+        segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.linalg.norm returns a ndarray, so when multiplied by field_vals, which
+    # is a YTArray, this leads to incorrect units. There does not appear to be
+    # a unyt implementation of norm, as far as I'm aware, so units will be
+    # handled manually for the time being
+    accum = np.cumsum(field_vals[:-1].d * np.linalg.norm(p[1:].d - p[:-1].d, axis=1))
+    accum = YTArray(accum, field_vals.units * p.units)
+    return accum
+
+
+class Accumulators:
+    r"""
+    Container for creating and storing the path integrals of various
+    fields along various paths.
+
+    The class takes in a list of user-defined paths and a dataset. From
+    these, the user can compute the integral of any field in the dataset
+    along the given paths. The results of these path integrals are
+    stored in a cumulative fashion, which means that, once computed, the
+    accumulated value of the field is stored at each point along the
+    path, allowing the user to query this information.
+    """
+
+    def __init__(self, paths, ds):
+        self.paths = []
+        self.ds = ds
+        self.ad = ds.all_data()
+        self.accum = []
+        self.left_edge = self.ds.domain_left_edge
+        self.right_edge = self.ds.domain_right_edge
+        # Make sure that the path has proper units. If no units are
+        # specified, assume code units
+        for p in paths:
```

Comment:
We need to make sure that paths, if it's a list of scalars, is converted to the right dimensionality.  I think the safest thing to do is to ensure (by validating the input) that `paths` is either a `YTArray` or a `np.ndarray`, and if it's a list, it needs to be a list of either scalars (in which case it gets dumped into a `(N,1)` array) or of ndarrays or ytarrays of all the same size.

I believe you can test this by checking if it's an array; if it is, then call `ensure_2d` on it.  If it's not -- if it's a list -- try converting it to a YTArray.